include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,323 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+#define BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+
+namespace boost {
+namespace urls {
+
+class decoded_view::const_iterator
+{
+    char const* begin_{nullptr};
+    char const* pos_{nullptr};
+    char const* end_{nullptr};
+    bool plus_to_space_{true};
+
+    friend decoded_view;
+
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space) noexcept
+        : begin_(str)
+        , pos_(str)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+    // end ctor
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space,
+        int) noexcept
+        : begin_(str)
+        , pos_(str + n)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+public:
+    using value_type = char;
+    using reference = char;
+    using pointer = void const*;
+    using const_reference = char;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    const_iterator() noexcept = default;
+
+    const_iterator(const_iterator const&) noexcept = default;
+
+    const_iterator&
+    operator=(const_iterator const&) noexcept = default;
+
+    reference
+    operator*() const noexcept
+    {
+        char c{'\0'};
+        pct_decode_opts opt;
+        opt.plus_to_space = plus_to_space_;
+        pct_decode_unchecked(
+            &c, &c+1, string_view(pos_, end_), opt);
+        return c;
+    }
+
+    const_iterator&
+    operator++() noexcept
+    {
+        BOOST_ASSERT(pos_ != nullptr);
+        pos_ += *pos_ == '%' ? 3 : 1;
+        return *this;
+    }
+
+    const_iterator&
+    operator--() noexcept
+    {
+        BOOST_ASSERT(pos_ != begin_);
+        if (pos_ - begin_ >= 3 &&
+            *(pos_ - 3) == '%')
+            pos_ -= 3;
+        else
+            --pos_;
+        return *this;
+    }
+
+    const_iterator
+    operator++(int) noexcept
+    {
+        auto tmp = *this;
+        ++*this;
+        return tmp;
+    }
+
+    const_iterator
+    operator--(int) noexcept
+    {
+        auto tmp = *this;
+        --*this;
+        return tmp;
+    }
+
+    const char*
+    base()
+    {
+        return pos_;
+    }
+
+    bool
+    operator==(
+        const_iterator const& other) const noexcept
+    {
+        return pos_ == other.pos_;
```

Comment:
I don't think this is enough. You have to also make sure that the iterators came from the same container. What if you have
```
string_view s = "deez nuts";
decoded_view d1( s );
decoded_view d2( string_view( s.data(), 4 ) );
```

now, iterators from d1 and d2 can compare equal when they shouldn't.

Reply:
> when they shouldn't

I'm not sure about that. Views in the STL return equal in this case: https://godbolt.org/z/56vxcdo4n

One might say that people making these decisions that are not smart, etc, and that we should not replicate these mistakes. But it's a recent standard and people are going to expect views of the same container to return equal.

Reply:
My intuition is that `std::string_view` has different design goals from us, and that we should not follow "make `decoded_view` work like other STL views" too closely. Our design goal is "optimize our container for our use-case." In light of this, functions like `base()` and `length()` are unnecessary. In fact, now that I think about this further, the equalty `pos_ == other.pos_` is OK, but we should add an assert that the iterators are from the same container. This will identify cases where users mistakenly attempt meaningless comparisons.

Perhaps no one will make this mistake, and it won't matter. Or, someone will want to actually perform the comparison using iterators from two different containers, and then open an issue about the assert. If this happens, we will then have an actual use-case that serves as a data point for analysis.

Reply:
This assumption (compared iterators come from the same container) is no different than the one already present for raw pointers. However, in the case of raw pointers `std::less` and `std::equal` guarantee that pointers from different containers can be compared. The use case is storing the iterators in containers.

Reply:
> The use case is storing the iterators in containers

Exactly. Which is effectively also a view.

Reply:
I agree with @akrzemi1 . In my mind, the `view` instance has no identity here (outside object identity). Instead it's like an iterator-pair. Different copies of the same iterator pair would (obviously) have iterators that compare equivalent too.

Reply:
(Also being able to compare iterators between subviews seems a useful feature)